docs/specs/tsd-006-startup-error-behavior.md

```diff
@@ -0,0 +1,105 @@
+# Startup Error Behavior
+
+## Objective
+
+Unified, configurable behavior on retry-able startup errors.
+
+## Keywords
+
+inputs, outputs, startup, error, retry
+
+## Overview
+
+Many Telegraf plugins connect to an external service either on the same machine
+or via network. On automated startup of Telegraf (e.g. via service) there is no
+guarantee that those services are fully started yet, especially when they reside
+on a remote host. More and more plugins implement mechanisms to retry reaching
+their related service if they failed to do so on startup.
+
+This specification intends to unify the naming of configuration-options, the
+values of those options and their semantic meaning. It describes the behavior
+for the different options on handling startup-errors.
+
+Startup-errors are all errors occurring in calls to `Start()`, for inputs and
+service-inputs, or `Connect()`, for outputs. The behaviors described below
+should only be applied in cases where the plugin *explicitly* states that an
+startup-error is *retry-able*. This comprises for example network errors
+indicating that the host or service is not reachable (yet) or external
+resources, like a machine or file, which are not yet available but might become
+available later. To indicate a retry-able startup-error the plugin should return
+a predefined error-type, potentially including an explicit flag. Additionally,
+the plugin might indicate partial connects to signal
+
+## Configuration Options and Behaviors
+
+Telegraf must introduce a unified `startup_error_behavior` configuration option
+for inputs and output plugins which is handled directly by the Telegraf
+infrastructure and is not passed down to the plugins. The setting must be
+available on a per-plugin basis and defines how Telegraf behaves on startup
+errors.
+
+For all listed behaviors Telegraf might retry to startup the plugin for a
+limited number of times during the startup phase before actually processing
+data.
+
+### `error` behavior
+
+If you set the `error` setting for the `startup_error_behavior` option Telegraf
+must fail and exit on startup-errors. This must be the default behavior.
+
+### `retry` behavior
+
+When using the `retry` setting for the `startup_error_behavior` option Telegraf
+must *not* fail on startup-errors and should continue running. Telegraf must
+retry to startup the failed plugin in each gather- or write-cycle, for inputs
+or for outputs respectively, for an unlimited number of times. Neither the
+plugin's `Gather()` nor `Write()` method is called as long as the startup did
+not succeed.
```

Comment:
Do we want to define the behavior of output buffers here as well?

Reply:
Hmmm we could, but then this spec also kind of scopes into how buffers are handled if no `Write()` is called. Should we do this here?

Explanation: If metrics are sent to the plugin and it is not yet connected, the metrics pile-up in the buffer until it finally connects.

Reply:
I would simply add your explanation + that metrics could start to overflow. How about that? I think what I was after is capturing what might happen in outputs as a consequence of this behavior.